File: proxy/python/utils/crypto.py
import json
import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from utils.keystore import load_or_create_key

_KEY: bytes = load_or_create_key()
_LEGACY_KEY = b"uxermanager-local-secret-key-32b"


def encrypt(text: str) -> str:
    iv = os.urandom(12)
    aesgcm = AESGCM(_KEY)
    ct_with_tag = aesgcm.encrypt(iv, text.encode("utf-8"), None)
    # AESGCM.encrypt returns ciphertext + 16-byte tag appended
    ct = ct_with_tag[:-16]
    tag = ct_with_tag[-16:]
    return json.dumps({
        "iv": iv.hex(),
        "tag": tag.hex(),
        "data": ct.hex(),
    })


def decrypt(encrypted_json: str) -> str:
    obj = json.loads(encrypted_json)
    iv = bytes.fromhex(obj["iv"])
    tag = bytes.fromhex(obj["tag"])
    ct = bytes.fromhex(obj["data"])
    aesgcm = AESGCM(_KEY)
    plaintext = aesgcm.decrypt(iv, ct + tag, None)
    return plaintext.decode("utf-8")


def decrypt_legacy(encrypted_json: str) -> str:
    obj = json.loads(encrypted_json)
    iv = bytes.fromhex(obj["iv"])
    tag = bytes.fromhex(obj["tag"])
    ct = bytes.fromhex(obj["data"])
    aesgcm = AESGCM(_LEGACY_KEY)
    plaintext = aesgcm.decrypt(iv, ct + tag, None)
    return plaintext.decode("utf-8")

File: proxy/nodejs/src/utils/crypto.js
const crypto = require('crypto');

const ALGORITHM = 'aes-256-gcm';
const { loadOrCreateKey } = require('./keystore');
let KEY;
try {
  KEY = loadOrCreateKey();
} catch (err) {
  throw new Error(`[UXERManager] 암호화 키 초기화 실패: ${err.message}`);
}
const LEGACY_KEY = Buffer.from('uxermanager-local-secret-key-32b', 'utf8');

function encrypt(text) {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv(ALGORITHM, KEY, iv);
  const encrypted = Buffer.concat([cipher.update(text, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return JSON.stringify({
    iv: iv.toString('hex'),
    tag: tag.toString('hex'),
    data: encrypted.toString('hex')
  });
}

function decrypt(encryptedJson) {
  const { iv, tag, data } = JSON.parse(encryptedJson);
  const decipher = crypto.createDecipheriv(ALGORITHM, KEY, Buffer.from(iv, 'hex'));
  decipher.setAuthTag(Buffer.from(tag, 'hex'));
  return decipher.update(Buffer.from(data, 'hex')) + decipher.final('utf8');
}

function decryptLegacy(encryptedJson) {
  const { iv, tag, data } = JSON.parse(encryptedJson);
  const decipher = crypto.createDecipheriv('aes-256-gcm', LEGACY_KEY, Buffer.from(iv, 'hex'));
  decipher.setAuthTag(Buffer.from(tag, 'hex'));
  return decipher.update(Buffer.from(data, 'hex')) + decipher.final('utf8');
}

module.exports = { encrypt, decrypt, decryptLegacy };
